File: EC_CTC.js
// CITY TRAFFIC CONTROL 

const express = require("express");
const bodyParser = require("body-parser");
const axios = require("axios");
const readline = require("readline");

const app = express();
app.use(bodyParser.json()); 

const API_KEY = "443334ba9843827f2569d98f613957b4"; 
let ciudad = "Alicante"; 

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

// Función para obtener la temperatura de la ciudad actual
const obtenerTemperatura = async () => {
  const url = `http://api.openweathermap.org/data/2.5/weather?q=${ciudad}&appid=${API_KEY}&units=metric`;
  try {
    const respuesta = await axios.get(url);
    if (respuesta.status === 200) {
      return respuesta.data.main.temp;
    } else {
      console.log(`Error al obtener temperatura: Código ${respuesta.status}`);
      return null;
    }
  } catch (error) {
    console.log("Error al conectar con OpenWeather:", error.message);
    throw new Error("Error al conectar con el servicio de OpenWeather.");
  }
};

// Función para verificar el estado de circulación
const verificarCirculacion = async () => {
  const temperatura = await obtenerTemperatura();
  if (temperatura !== null) {
    return {
      temperatura,
      estado: temperatura >= 0 ? "OK" : "KO",
    };
  }
  return { temperatura: null, estado: "Error al obtener la temperatura" };
};

// Función para mostrar el menú interactivo
const mostrarMenu = () => {
  console.log("\n--- Menú Interactivo City Traffic Control ---");
  console.log(`Ciudad actual: ${ciudad}`);
  console.log("1. Verificar estado de circulación");
  console.log("2. Cambiar ciudad");
  console.log("3. Salir");

  rl.question("Seleccione una opción: ", async (opcion) => {
    switch (opcion) {
      case "1":
        try {
          const { temperatura, estado } = await verificarCirculacion();
          console.log(
            `\nTemperatura actual en ${ciudad}: ${temperatura}°C\nEstado de circulación en ${ciudad}: ${estado}`
          );
        } catch (error) {
          console.log("Error al verificar el estado de circulación:", error.message);
        }
        mostrarMenu(); 
        break;
      case "2":
        rl.question("\nIngrese el nombre de la nueva ciudad: ", (nuevaCiudad) => {
          ciudad = nuevaCiudad;
          console.log(`Ciudad cambiada a: ${ciudad}`);
          mostrarMenu(); 
        });
        break;
      case "3":
        console.log("\nSaliendo del menú interactivo...\n");
        rl.close();
        process.exit(0); 
        break;
      default:
        console.log("\nOpción no válida. Intente de nuevo.\n");
        mostrarMenu(); 
    }
  });
};

// Función para emular el comportamiento del menú desde REST
const procesarAccionRestComoMenu = async (opcion) => {
  switch (opcion) {
    case "1":
      const { temperatura, estado } = await verificarCirculacion();
      console.log(
        `\n\n[REST] Estado de circulación solicitado desde REST:\nTemperatura actual en ${ciudad}: ${temperatura}°C\nEstado de circulación en ${ciudad}: ${estado}`
      );
      break;
    case "2":
      console.log(`\n[REST] Ciudad cambiada desde REST:\nCiudad cambiada a: ${ciudad}`);
      break;
    default:
      console.log("\nAcción no válida desde REST.\n");
  }
  mostrarMenu(); 
};

// Endpoint para obtener el estado de circulación
app.get("/estado_circulacion", async (req, res, next) => {
  try {
    const { temperatura, estado } = await verificarCirculacion();
    await procesarAccionRestComoMenu("1"); 
    res.json({
      ciudad: ciudad,
      estado: estado,
      temperatura: temperatura !== null ? `${temperatura}°C` : "No disponible",
    });
  } catch (error) {
    next(error); 
  }
});

// Endpoint para cambiar la ciudad
app.post("/cambiar_ciudad", (req, res, next) => {
  try {
    const { nuevaCiudad } = req.body;
    if (!nuevaCiudad) {
      return res.status(400).json({ error: "El parámetro 'nuevaCiudad' es requerido." });
    }
    ciudad = nuevaCiudad;
    console.log(`\n[REST] Ciudad cambiada desde REST:`);
    procesarAccionRestComoMenu("2"); // Emular la acción en el menú
    res.json({ mensaje: `Ciudad cambiada a ${ciudad}` });
  } catch (error) {
    next(error);
  }
});

// Middleware para manejo global de errores
app.use((err, req, res, next) => {
  console.error("Error inesperado:", err.message || err);
  res.status(500).json({
    error: "Ha ocurrido un error inesperado. Por favor, inténtalo de nuevo más tarde.",
  });
});

// Middleware para manejar rutas no encontradas
app.use((req, res) => {
  res.status(404).json({
    error: "Ruta no encontrada. Verifica el endpoint solicitado.",
  });
});

// Iniciar el servidor
const PORT = 3000;
app.listen(PORT, () => {
  console.log(`Servidor REST ejecutándose en http://localhost:${PORT}`);
  console.log("\nPara usar la API REST de City Traffic Control:");
  console.log("- GET /estado_circulacion para verificar estado de circulación");
  console.log("- POST /cambiar_ciudad con JSON { 'nuevaCiudad': 'nombre' } para cambiar de ciudad");
  console.log("\nTambién se puede usar el Menú Interactivo:");
  mostrarMenu(); 
});

File: EC_CTC.py
import requests
from flask import Flask, jsonify, request

API_KEY = "443334ba9843827f2569d98f613957b4"  
ciudad = "Alicante"

# Función para cambiar de ciudad
def cambiar_ciudad(nueva_ciudad):
    global ciudad
    ciudad = nueva_ciudad
    print(f"Ciudad cambiada a: {ciudad}")

# Función para obtener la temperatura de la ciudad actual
def obtener_temperatura():
    global ciudad, API_KEY
    url = f"http://api.openweathermap.org/data/2.5/weather?q={ciudad}&appid={API_KEY}&units=metric"
    try:
        respuesta = requests.get(url)
        if respuesta.status_code == 200:
            datos = respuesta.json()
            temperatura = datos['main']['temp']
            return temperatura
        elif respuesta.status_code == 404:
            print("Ciudad no encontrada. Verifica el nombre.")
            return None
        elif respuesta.status_code == 401:
            print("Error de autenticación. Verifica tu clave API.")
            return None
        else:
            print(f"Error al obtener datos del clima. Código: {respuesta.status_code}")
            return None
    except Exception as e:
        print(f"Error al conectarse a la API: {e}")
        return None

# Función para verificar el estado de circulación
def verificar_circulacion():
    temperatura = obtener_temperatura()
    estado = None
    if temperatura is not None:
        if temperatura >= 0:
            estado = "OK"
        else:
            estado = "KO"
    return temperatura, estado

# Menú interactivo
def mostrar_menu():
    while True:
        print("\n--- Control de Tráfico Urbano ---")
        print(f"Ciudad actual: {ciudad}")
        print("1. Verificar estado de circulación")
        print("2. Cambiar ciudad")
        print("3. Salir")
        opcion = input("Seleccione una opción: ")

        if opcion == "1":
            temperatura, estado = verificar_circulacion()
            print(f"Temperatura actual: {temperatura}°C")
            print(f"Estado de circulación en {ciudad}: {estado}")
        elif opcion == "2":
            nueva_ciudad = input("Ingrese el nombre de la nueva ciudad: ")
            cambiar_ciudad(nueva_ciudad)
        elif opcion == "3":
            print("Saliendo...")
            break
        else:
            print("Opción no válida. Intente de nuevo.")

# Configuración de API REST con Flask
app = Flask(__name__)

from flask import Flask, jsonify, request, json

@app.route("/estado_circulacion", methods=["GET"])
def api_estado_circulacion():
    global ciudad
    temperatura, estado = verificar_circulacion()
    print(f"Temperatura actual: {temperatura}°C")
    print(f"Estado de circulación en {ciudad}: {estado}")    
    if temperatura is not None:
        return app.response_class(
            response=json.dumps({
                "ciudad": ciudad,
                "estado": estado,
                "temperatura": f"{temperatura} °C"
            }, ensure_ascii=False),  
            status=200,
            mimetype="application/json"
        )
    else:
        return app.response_class(
            response=json.dumps({
                "ciudad": ciudad,
                "estado": estado,
                "temperatura": "No disponible"
            }, ensure_ascii=False),  
            status=200,
            mimetype="application/json"
        )

@app.route("/cambiar_ciudad", methods=["POST"])
def api_cambiar_ciudad():
    global ciudad
    datos = request.json  
    if "ciudad" in datos:
        nueva_ciudad = datos["ciudad"]
        cambiar_ciudad(nueva_ciudad)
        return jsonify({"mensaje": f"Ciudad cambiada a {nueva_ciudad}", "ciudad": ciudad})
    else:
        return jsonify({"error": "El parámetro 'ciudad' es requerido"}), 400

# Main
if __name__ == "__main__":
    print("Inicio del sistema City Traffic Control")
    print(f"Ciudad predeterminada: {ciudad}")
    
    # Se puede elegir entre menú interactivo o API REST
    modo = input("¿Menú interactivo (1) o servidor REST (2)? ")
    
    if modo == "1":
        mostrar_menu()
    elif modo == "2":
        print("Para observar el estado de circulación, accede a http://127.0.0.1:5000/estado_circulacion")
        print("Para cambiar de ciudad, usa una solicitud POST a http://127.0.0.1:5000/cambiar_ciudad")
        app.run(port=5000)
    else:
        print("Opción no válida. Saliendo...")
